chrome/content/zotero/xpcom/reader.js

```diff
@@ -555,35 +568,41 @@ class ReaderInstance {
 					this._resolveInitPromise();
 					return;
 				}
-				case 'setAnnotation': {
+				case 'saveAnnotations': {
 					let attachment = Zotero.Items.get(data.itemID);
-					let { annotation } = message;
-					annotation.key = annotation.id;
-					let saveOptions = {
-						notifierData: {
-							instanceID: this._instanceID
+					let { annotations } = message;
+					let notifierQueue = new Zotero.Notifier.Queue();
+					for (let annotation of annotations) {
+						annotation.key = annotation.id;
+						let saveOptions = {
+							notifierQueue,
+							notifierData: {
+								instanceID: this._instanceID
+							}
+						};
+						let savedAnnotation = await Zotero.Annotations.saveFromJSON(attachment, annotation, saveOptions);
+						if (annotation.image && !await Zotero.Annotations.hasCacheImage(savedAnnotation)) {
+							let blob = this._dataURLtoBlob(annotation.image);
+							await Zotero.Annotations.saveCacheImage(savedAnnotation, blob);
 						}
-					};
-					let savedAnnotation = await Zotero.Annotations.saveFromJSON(attachment, annotation, saveOptions);
-					
-					if (annotation.image && !await Zotero.Annotations.hasCacheImage(savedAnnotation)) {
-						let blob = this._dataURLtoBlob(annotation.image);
-						await Zotero.Annotations.saveCacheImage(savedAnnotation, blob);
 					}
+					await Zotero.Notifier.commit(notifierQueue);
```

Comment:
This (and others) should be in a `finally`, or else annotations that are saved successfully won't be updated properly and the UI will get out of sync.